Cabal-syntax/src/Distribution/Types/BuildInfo.hs

```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE DataKinds #-}
```

Comment:
The use of this extension is, I believe, entirely due to changing the design of the symbolic path datatype. It used to be that you declared individual empty data-types:

```haskell
data PackageDir
data SourceDir
data SourceFile
instance IsDir PackageDir
instance IsDir SourceDir
```

and then you would have

```haskell
type SymbolicPath :: Type -> Type -> Type

hsSourceDirs :: [ SymbolicPath PackageDir SourceDir ]
```

I didn't like the typeclass mechanism using `IsDir`; I think it makes a lot more sense to distinguish directly in the type whether something is a file or a directory instead of relying on a typeclass for which instances could get added later on. So I then went for:

```haskell
data FileOrDir a = File a | Dir a
type SymbolicPath :: a -> FileOrDir a -> Type

hsSourceDirs :: [ SymbolicPath PackageDir (Dir SourceDir) ]
```

This gets rid of the instances, but it does use `DataKinds`.

It was still a bit of a chore to define all these empty data-types, so I changed it further to use type-level strings:

```haskell
data FileOrDir = File Symbolc | Dir Symbol
type SymbolicPath :: Symbol -> FileOrDir -> Type

hsSourceDirs :: [ SymbolicPath "Package" (Dir "Source") ]
```

Now, it can be argued that it is better to have datatypes rather than raw strings: a datatype is properly namespaced by the module that defines it, and has to be imported in order to be referred to. This avoids issues in which two different modules define a different type of abstraction location with the same symbolic name. Given the amount of `SymbolicPath`s now in the codebase, it lead to a proliferation of data type declarations that in my opinion bloated the patch. In particular, it made it hard to iterate on the patch as each time I came across a new kind of abstract filepath I had to find in which module it made sense to define a corresponding new datatype, and then had to change import/export lists to include it.

Reply:
I'm also using `DataKinds` to distinguish symbolic paths that must be relative from symbolic paths that are allowed to be absolute:

```haskell
-- | Is this symbolic path allowed to be absolute, or must it be relative?
data AllowAbsolute
  = -- | The path may be absolute, or it may be relative.
    AllowAbsolute
  | -- | The path must be relative.
    OnlyRelative
```

This allows functions to be polymorphic about whether paths must be absolute or relative, for example:

```haskell
findFileCwd
  :: forall dir1 dir2 file allowAbsolute
   . Verbosity
  -> Maybe (SymbolicPath "CWD" (Dir dir1))
  -- ^ working directory
  -> [SymbolicPathX allowAbsolute dir1 (Dir dir2)]
  -- ^ search directories
  -> RelativePath dir2 (File file)
  -- ^ File Name
  -> IO (SymbolicPathX allowAbsolute dir1 (File file))
  ```
  
If `findFileCwd` gets passed in search directories that are guaranteed to be relative, it will return file paths that are themselves relative. Without `DataKinds`, I think I would end up having to define two separate datatypes and then a function such as this would have to inherit a class constraint in order to retain polymorphism.

Reply:
Don't worry about.  I'll write my own cabal replacement.  The source for cabal is too far gone. :)